rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+
+**OCaml**
+
+| | Platform | Task | Effort | Due | Notes |
+| -- | ----------- | ----------- | ------ | ---------- | ---------------|
+| 10 | all | OCaml/Dune learning curve | 4 | ongoing |  I can usually fake it, since all languages are alike, but darnit, sometimes I have to sit down and actually learn something. |
+| 11 | all | Analyze existing Bazel/OCaml libs | 4 | ongoing | |
+| 12 | all | Develop rules_ocaml | 12 |Sun 5/31 |
+| 13 | all | Implement Bazel support for Coda OCaml code | 16 | Sun 5/31 |
+
+**Golang, CUDA**
+
+| | Platform | Task | Effort | Due | Notes |
+| -- | ----------- | ----------- | ------ | ---------- | ---------------|
+| 14 | all | go-libp2p | 4 | W 5/27  |
+| 15 | all | CUDA code in Coda | 1 | TBD | cuda-fixnum, cuda-prover, etc. |
+
+Note: I don't have a GPU, and I'm not sure if I can compile CUDA code
+without one.  If I can, task 15 will be completed by end of month.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The learning curve for Bazel is little steep in places, plus it has
+its own languge, Starlark - yet another language to learn.  It's a
+Python dialect, but still, somebody on staff will probably have to be
+tasked with learning the stuff.  On the other hand, once the build
+system is in place it will probably not require much further work,
+just like Makefiles are usually left alone once they stabilize.
+
+Another issue with Bazel is feature detection.  Bazel just builds
+stuff; it does not provide the feature-detection functionality
+supported by GNU autoconf tools.  So it has no analog to
+`./configure`; you have to explicitly tell it about all features.
+That means that a Bazel build, in cases where feature detection is
+required, must be complimented by a traditional `configure` file or
+something similar.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Descriptions of the general benefits of Bazel are easily found with a
+simple websearch, so I'll keep it brief:
+* Reliable, replicable builds.
+* Support for multiple languages.
+* Developed and heavily used by Google; also used by many other major projects.
+* Eliminates the need to separately install/config/build third-party
+  dependencies (apt-get, homebrew, etc.) Bazel will download and build
+  everything.
+* Support for remote caching and execution is (probably) unique.
+
+Alternatives
+* Stick with current situation.  That might work for o(1) internally
+  but is not so attractive for outside volunteers who would have to
+  struggle with the build system.  My guess is that improving the
+  current system would probably be more expensive in the long run than
+  switching to Bazel.
+* Nix may be an alternative, but according to the folks who provide
+  Bazel support for Haskell, Bazel and Nix work well together: [NIX +
+  BAZEL = FULLY REPRODUCIBLE, INCREMENTAL
+  BUILDS](https://www.tweag.io/posts/2018-03-15-bazel-nix.html).
```

Comment:
Yes, the Nix/Bazel hybrid would work really well. ie. Bazel for the core project build, Nix for the deps on the outsides. We did this at Pinterest for our iOS builds and the combo worked very well. Of course, keep this part out of scope for now.